File: hello-typescript/src/server/memory/MemoryMappedCache.ts
/**
 * Memory-mapped cache for audio streaming.
 * Provides file-based caching with memory mapping.
 * Matches C++ MemoryMappedCache and Java MemoryMappedCache.
 * Follows the unified mmap implementation specification v2.0.0.
 */

import * as fs from "fs";
import mmap from "mmap-io";

// Configuration constants - follows unified mmap implementation specification v2.0.0
const DEFAULT_PAGE_SIZE = 64 * 1024 * 1024; // 64MB
const MAX_CACHE_SIZE = 8 * 1024 * 1024 * 1024; // 8GB
const SEGMENT_SIZE = 1 * 1024 * 1024 * 1024; // 1GB per segment
const BATCH_OPERATION_LIMIT = 1000; // Max batch operations

export class MemoryMappedCache {
  private path: string;
  private fd: number | null;
  private size: number;
  private isOpenFlag: boolean;
  private buffer: Buffer | null;

  constructor(path: string) {
    this.path = path;
    this.fd = null;
    this.size = 0;
    this.isOpenFlag = false;
    this.buffer = null;
  }

  create(initialSize: number = 0): boolean {
    try {
      if (fs.existsSync(this.path)) {
        fs.unlinkSync(this.path);
      }

      this.fd = fs.openSync(this.path, "w+");

      if (initialSize > 0) {
        fs.ftruncateSync(this.fd, initialSize);
        this.size = initialSize;
        this.mapFile();
      } else {
        this.size = 0;
      }

      this.isOpenFlag = true;
      console.log(`Created mmap file: ${this.path} with size: ${initialSize}`);
      return true;
    } catch (err: any) {
      console.error(`Error creating file ${this.path}: ${err.message}`);
      return false;
    }
  }

  open(): boolean {
    try {
      if (!fs.existsSync(this.path)) {
        console.error(`File does not exist: ${this.path}`);
        return false;
      }

      this.fd = fs.openSync(this.path, "r+");
      const stats = fs.fstatSync(this.fd);
      this.size = stats.size;

      if (this.size > 0) {
        this.mapFile();
      }

      this.isOpenFlag = true;
      console.log(`Opened mmap file: ${this.path} with size: ${this.size}`);
      return true;
    } catch (err: any) {
      console.error(`Error opening file ${this.path}: ${err.message}`);
      return false;
    }
  }

  close(): void {
    if (this.isOpenFlag && this.fd !== null) {
      try {
        if (this.buffer) {
          // Unmap if mapped
          // mmap-io cleanup relying on GC or closing fd
        }
        fs.closeSync(this.fd);
        this.fd = null;
        this.isOpenFlag = false;
        this.buffer = null;
      } catch (err: any) {
        console.error(`Error closing file: ${err.message}`);
      }
    }
  }

  write(offset: number, data: Buffer): number {
    if (!this.isOpenFlag || this.fd === null) {
      const initialSize = offset + data.length;
      if (!this.create(initialSize)) {
        return 0;
      }
    }

    const requiredSize = offset + data.length;
    if (requiredSize > this.size) {
      if (!this.resize(requiredSize)) {
        console.error("Failed to resize file for write operation");
        return 0;
      }
    }

    try {
      // Write to memory buffer
      if (this.buffer) {
        data.copy(this.buffer, offset);
        return data.length;
      } else {
        // Fallback if buffer not mapped (should not happen if logic correct)
        fs.writeSync(this.fd!, data, 0, data.length, offset);
        return data.length;
      }
    } catch (err: any) {
      console.error(`Error writing to file ${this.path}: ${err.message}`);
      return 0;
    }
  }

  read(offset: number, length: number): Buffer {
    if (!this.isOpenFlag || this.fd === null) {
      if (!this.open()) {
        console.error(`Failed to open file for reading: ${this.path}`);
        return Buffer.alloc(0);
      }
    }

    if (offset >= this.size) {
      return Buffer.alloc(0);
    }

    const actualLength = Math.min(length, this.size - offset);

    try {
      if (this.buffer) {
        const result = Buffer.alloc(actualLength);
        this.buffer.copy(result, 0, offset, offset + actualLength);
        console.log(`Read ${actualLength} bytes from ${this.path} at offset ${offset}`);
        return result;
      } else {
        const buffer = Buffer.alloc(actualLength);
        const bytesRead = fs.readSync(this.fd!, buffer, 0, actualLength, offset);
        console.log(`Read ${bytesRead} bytes from ${this.path} at offset ${offset}`);
        return buffer;
      }
    } catch (err: any) {
      console.error(`Error reading from file ${this.path}: ${err.message}`);
      return Buffer.alloc(0);
    }
  }

  resize(newSize: number): boolean {
    if (!this.isOpenFlag) {
      console.error(`File not open for resize: ${this.path}`);
      return false;
    }

    if (newSize === this.size) {
      return true;
    }

    try {
      fs.ftruncateSync(this.fd!, newSize);
      this.size = newSize;

      // Re-map with new size
      this.mapFile();

      console.log(`Resized file ${this.path} to ${newSize} bytes`);
      return true;
    } catch (err: any) {
      console.error(`Error resizing file ${this.path}: ${err.message}`);
      return false;
    }
  }

  flush(): boolean {
    if (!this.isOpenFlag || this.fd === null) {
      console.warn(`File not open for flush: ${this.path}`);
      return false;
    }

    try {
      if (this.buffer) {
        // mmap-io sync: mmap.sync(buffer, offset, length, blocking_sync, invalidate_pages)
        mmap.sync(this.buffer, 0, this.size, true);
      } else {
        fs.fsyncSync(this.fd);
      }

      console.log(`Flushed file: ${this.path}`);
      return true;
    } catch (err: any) {
      console.error(`Error flushing file: ${err.message}`);
      return false;
    }
  }

  finalize(finalSize: number): boolean {
    if (!this.isOpenFlag) {
      console.warn(`File not open for finalization: ${this.path}`);
      return false;
    }

    if (!this.resize(finalSize)) {
      console.error(`Failed to resize file during finalization: ${this.path}`);
      return false;
    }

    // Sync to disk
    this.flush();

    console.log(`Finalized file: ${this.path} with size: ${finalSize}`);
    return true;
  }

  getSize(): number {
    return this.size;
  }

  getPath(): string {
    return this.path;
  }

  isOpen(): boolean {
    return this.isOpenFlag;
  }

  private mapFile(): void {
    if (this.fd !== null && this.size > 0) {
      // mmap.map(size, protection, privacy, fd, offset, advise)
      // PROT_READ | PROT_WRITE, MAP_SHARED
      this.buffer = mmap.map(
        this.size,
        (mmap.PROT_READ | mmap.PROT_WRITE) as any,
        mmap.MAP_SHARED,
        this.fd,
        0
      );
    }
  }
}

File: hello-rust/src/server/memory/memory_mapped_cache.rs
// Memory-mapped cache for efficient file I/O.
// Provides write, read, resize, and finalize operations.
// Matches Python MmapCache functionality.

use memmap2::MmapMut;
use std::fs::{File, OpenOptions};
use std::path::Path;
use std::sync::Mutex;

// Configuration constants - follows unified mmap specification v2.0.0
#[allow(dead_code)]
const DEFAULT_PAGE_SIZE: u64 = 64 * 1024 * 1024; // 64MB
#[allow(dead_code)]
const MAX_CACHE_SIZE: u64 = 8 * 1024 * 1024 * 1024; // 8GB
#[allow(dead_code)]
const SEGMENT_SIZE: u64 = 1 * 1024 * 1024 * 1024; // 1GB per segment
#[allow(dead_code)]
const BATCH_OPERATION_LIMIT: usize = 1000; // Max batch operations

/// Memory-mapped cache implementation using memmap2.
#[allow(dead_code)]
pub struct MemoryMappedCache {
    path: String,
    file: Mutex<Option<File>>,
    mmap: Mutex<Option<MmapMut>>,
    size: Mutex<u64>,
    is_open: Mutex<bool>,
}

#[allow(dead_code)]
impl MemoryMappedCache {
    /// Create a new MemoryMappedCache.
    pub fn new(path: String) -> Self {
        Self {
            path,
            file: Mutex::new(None),
            mmap: Mutex::new(None),
            size: Mutex::new(0),
            is_open: Mutex::new(false),
        }
    }

    /// Create a new memory-mapped file.
    pub fn create(&self, initial_size: u64) -> bool {
        let mut file_lock = self.file.lock().unwrap();

        // Remove existing file
        if Path::new(&self.path).exists() {
            let _ = std::fs::remove_file(&self.path);
        }

        // Create and open file
        let file = match OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .open(&self.path)
        {
            Ok(f) => f,
            Err(e) => {
                eprintln!("Error creating file {}: {:?}", self.path, e);
                return false;
            }
        };

        // Set file size
        if initial_size > 0 {
            if let Err(e) = file.set_len(initial_size) {
                eprintln!("Error truncating file {}: {:?}", self.path, e);
                return false;
            }
        }

        *self.size.lock().unwrap() = initial_size;
        *file_lock = Some(file);

        // Map file into memory if size > 0
        if initial_size > 0 {
            if !self.map_file() {
                return false;
            }
        }

        *self.is_open.lock().unwrap() = true;
        println!(
            "Created mmap file: {} with size: {}",
            self.path, initial_size
        );
        true
    }

    /// Open an existing memory-mapped file.
    pub fn open(&self) -> bool {
        if !Path::new(&self.path).exists() {
            eprintln!("File does not exist: {}", self.path);
            return false;
        }

        let file = match OpenOptions::new().read(true).write(true).open(&self.path) {
            Ok(f) => f,
            Err(e) => {
                eprintln!("Error opening file {}: {:?}", self.path, e);
                return false;
            }
        };

        let size = match file.metadata() {
            Ok(metadata) => metadata.len(),
            Err(e) => {
                eprintln!("Error getting file stats for {}: {:?}", self.path, e);
                return false;
            }
        };

        *self.size.lock().unwrap() = size;
        *self.file.lock().unwrap() = Some(file);

        // Map file into memory if size > 0
        if size > 0 {
            if !self.map_file() {
                return false;
            }
        }

        *self.is_open.lock().unwrap() = true;
        println!("Opened mmap file: {} with size: {}", self.path, size);
        true
    }

    /// Close memory-mapped file.
    pub fn close(&self) {
        if *self.is_open.lock().unwrap() {
            self.unmap_file();
        }
    }

    /// Write data to memory-mapped file.
    pub fn write(&self, offset: u64, data: &[u8]) -> usize {
        // Check if file is open
        if !*self.is_open.lock().unwrap() {
            let initial_size = offset + data.len() as u64;
            if !self.create(initial_size) {
                return 0;
            }
        }

        // Check required size
        let required_size = offset + data.len() as u64;
        let current_size = *self.size.lock().unwrap();
        let has_mmap = self.mmap.lock().unwrap().is_some();

        // If file needs to grow or has no mmap yet, resize it
        if required_size > current_size || !has_mmap {
            let new_size = std::cmp::max(required_size, current_size);
            if !self.resize(new_size) {
                eprintln!("Failed to resize file for write operation");
                return 0;
            }
        }

        let mut mmap_lock = self.mmap.lock().unwrap();
        if let Some(ref mut mmap) = *mmap_lock {
            let offset = offset as usize;
            if offset + data.len() <= mmap.len() {
                mmap[offset..offset + data.len()].copy_from_slice(data);
                println!(
                    "Wrote {} bytes to {} at offset {}",
                    data.len(),
                    self.path,
                    offset
                );
                data.len()
            } else {
                eprintln!("Write offset out of bounds");
                0
            }
        } else {
            eprintln!("No mmap available after resize");
            0
        }
    }

    /// Read data from memory-mapped file.
    pub fn read(&self, offset: u64, length: usize) -> Vec<u8> {
        if !*self.is_open.lock().unwrap() || self.mmap.lock().unwrap().is_none() {
            if !self.open() {
                eprintln!("Failed to open file for reading: {}", self.path);
                return Vec::new();
            }
        }

        let size = *self.size.lock().unwrap();
        if offset >= size {
            return Vec::new();
        }

        let actual_length = std::cmp::min(length, (size - offset) as usize);

        let mmap_lock = self.mmap.lock().unwrap();
        if let Some(ref mmap) = *mmap_lock {
            let offset = offset as usize;
            if offset + actual_length <= mmap.len() {
                let data = mmap[offset..offset + actual_length].to_vec();
                println!(
                    "Read {} bytes from {} at offset {}",
                    data.len(),
                    self.path,
                    offset
                );
                data
            } else {
                eprintln!("Read offset out of bounds");
                Vec::new()
            }
        } else {
            Vec::new()
        }
    }

    /// Get the size of the file.
    pub fn get_size(&self) -> u64 {
        *self.size.lock().unwrap()
    }

    /// Get the path of the file.
    pub fn get_path(&self) -> &str {
        &self.path
    }

    /// Check if the file is open.
    pub fn is_open(&self) -> bool {
        *self.is_open.lock().unwrap()
    }

    /// Resize the file to a new size.
    pub fn resize(&self, new_size: u64) -> bool {
        if !*self.is_open.lock().unwrap() {
            eprintln!("File not open for resize: {}", self.path);
            return false;
        }

        // Unmap current mmap
        self.unmap_file();

        // Resize file
        {
            let mut file_lock = self.file.lock().unwrap();
            if let Some(ref mut file) = *file_lock {
                if let Err(e) = file.set_len(new_size) {
                    eprintln!("Error resizing file {}: {:?}", self.path, e);
                    return false;
                }
            }
        } // Release file_lock here

        *self.size.lock().unwrap() = new_size;

        // Remap file if size > 0
        if new_size > 0 {
            if !self.map_file() {
                return false;
            }
        }

        println!("Resized file {} to {} bytes", self.path, new_size);
        true
    }

    /// Flush all mapped data to disk.
    pub fn flush(&self) -> bool {
        if !*self.is_open.lock().unwrap() {
            eprintln!("File not open for flush: {}", self.path);
            return false;
        }

        if let Some(ref mmap) = *self.mmap.lock().unwrap() {
            if let Err(e) = mmap.flush() {
                eprintln!("Error flushing file {}: {:?}", self.path, e);
                return false;
            }
        }

        println!("Flushed file: {}", self.path);
        true
    }

    /// Finalize the file to its final size.
    pub fn finalize(&self, final_size: u64) -> bool {
        if !*self.is_open.lock().unwrap() {
            eprintln!("File not open for finalization: {}", self.path);
            return false;
        }

        if !self.resize(final_size) {
            eprintln!("Failed to resize file during finalization: {}", self.path);
            return false;
        }

        // MmapMut flushes automatically when dropped, but we can force flush
        if let Some(ref mmap) = *self.mmap.lock().unwrap() {
            mmap.flush().ok();
        }

        println!("Finalized file: {} with size: {}", self.path, final_size);
        true
    }

    /// Map the file into memory using memmap2.
    fn map_file(&self) -> bool {
        let file_lock = self.file.lock().unwrap();
        if let Some(ref file) = *file_lock {
            let size = *self.size.lock().unwrap() as usize;
            if size > 0 {
                // Map entire file into memory (read-write mode)
                match unsafe { MmapMut::map_mut(file) } {
                    Ok(mmap) => {
                        *self.mmap.lock().unwrap() = Some(mmap);
                        println!("Successfully mapped file: {} ({} bytes)", self.path, size);
                        true
                    }
                    Err(e) => {
                        eprintln!("Error mapping file {}: {:?}", self.path, e);
                        false
                    }
                }
            } else {
                false
            }
        } else {
            false
        }
    }

    /// Unmap the file from memory.
    fn unmap_file(&self) {
        *self.mmap.lock().unwrap() = None;
    }
}
